### ISSUE TITLE
Add hints to challenges

### ISSUE THREAD

**Issue by Mesure73L** [Mesure73L/mesure.x10.mx#111]
### What problem were you having?

Some challenges will definitely be harder than others. The problem with this is Twelve of Code is supposed to be for all developers, and some challenges might be too hard for them resulting in them giving up.

### What solution would you like?

Add hints to challenges. These hints could either release with the challenges, or maybe 5 days (or something like that) later, to give time for the developers to think of a solution and then if they are stuck they could get a hint.

### What alternatives have you considered?

An alternative I've thought of is maybe adding a way to skip a challenge, which will not show as completed but will allow you to have a year/month marked as completed without finishing every single challenges.

Or we could do both of these!

#### Pros:

- Gives newer developers a way to get a little bit of help without spoiling the solution
- Allows newer developers to finish the challenges instead of giving up

#### Cons:

- People might misuse the hints and read them without trying the challenge first
- I don't know a good spot to put it

### Solution to the First Con

I thought about this, and maybe you have to have *x* failed attempts before you can get a hint. But then people might purposefully type random gibberish to get the hint.

### Other Note

What milestone should I put this in, if it's going to happen? I'm not sure if we're gonna add it to 1.0.0 or 1.1.0, so which (if any) should I put it in?

### What page is this for?

Twelve of Code

**Comment by BeepBot99** [comment]
> Add hints to challenges. These hints could either release with the challenges, or maybe 5 days (or something like that) later, to give time for the developers to think of a solution and then if they are stuck they could get a hint.

Yes, 5 days sounds good.

> An alternative I've thought of is maybe adding a way to skip a challenge, which will not show as completed but will allow you to have a year/month marked as completed without finishing every single challenges.

No, I don't like that idea.

> * People might misuse the hints and read them without trying the challenge first
> * I don't know a good spot to put it

Sure!

> I thought about this, and maybe you have to have _x_ failed attempts before you can get a hint. But then people might purposefully type random gibberish to get the hint.

No, don't do that.
**REMEMBER:**
12oC is just a personal challenge, it's not competitive. It doesn't matter. If they cheat, then all that will happen is they'll feel guilty later.

> What milestone should I put this in, if it's going to happen? I'm not sure if we're gonna add it to 1.0.0 or 1.1.0, so which (if any) should I put it in?

How about 1.0.0.

**Comment by BeepBot99** [comment]
Low priority though.

**Comment by BeepBot99** [comment]
I'm moving this to backlog because we'll only do it if we have time to do it.

**Comment by Mesure73L** [comment]
> > Add hints to challenges. These hints could either release with the challenges, or maybe 5 days (or something like that) later, to give time for the developers to think of a solution and then if they are stuck they could get a hint.
> 
> Yes, 5 days sounds good.

K, no comment.

> > An alternative I've thought of is maybe adding a way to skip a challenge, which will not show as completed but will allow you to have a year/month marked as completed without finishing every single challenges.
> 
> No, I don't like that idea.

Why not? Why don't you like that idea?

> > * People might misuse the hints and read them without trying the challenge first
> > * I don't know a good spot to put it
> 
> Sure!

What do you mean by `Sure!`?

> > I thought about this, and maybe you have to have _x_ failed attempts before you can get a hint. But then people might purposefully type random gibberish to get the hint.
> 
> No, don't do that. **REMEMBER:** 12oC is just a personal challenge, it's not competitive. It doesn't matter. If they cheat, then all that will happen is they'll feel guilty later.

Yes, 12oC is not competitive, but I still want to push people to doing it legitimately or else it'll just get stale and they won't actually try.

> > What milestone should I put this in, if it's going to happen? I'm not sure if we're gonna add it to 1.0.0 or 1.1.0, so which (if any) should I put it in?
> 
> How about 1.0.0.

Sure, no comment

**Comment by BeepBot99** [comment]
> Why not? Why don't you like that idea?

I think people can either do them or not do them, not lie about doing them. Plus, why does it matter if they can see the month or year as completed? Why would they care?

> What do you mean by `Sure!`?
> 

Have you never seen `Sure!` used like that before? It means, `Totally!`.

> Yes, 12oC is not competitive, but I still want to push people to doing it legitimately or else it'll just get stale and they won't actually try.

So, how about 5 days then. There could maybe also be a place to see how long it took, so it would show that it took you five days. It could also show if you used the hint or not.

**Comment by Mesure73L** [comment]
> I think people can either do them or not do them, not lie about doing them. Plus, why does it matter if they can see the month or year as completed? Why would they care?

Some people might want to see it as completed, **you** might not care but some people probably will

> Have you never seen `Sure!` used like that before? It means, `Totally!`.

I DIDNT ASK A QUESTION
I KNOW WHAT `Sure!` MEANS, JUST NOT USED IN THIS CONTEXT WHERE **I DIDNT ASK YOU ANY QUESTIONS AND IT MAKES NO SENSE**

> So, how about 5 days then. There could maybe also be a place to see how long it took, so it would show that it took you five days. It could also show if you used the hint or not.

True

**Comment by BeepBot99** [comment]
> > I think people can either do them or not do them, not lie about doing them. Plus, why does it matter if they can see the month or year as completed? Why would they care?
> 
> Some people might want to see it as completed, **you** might not care but some people probably will
> 

No, either someone can actually complete it, not complete it, or manipulate `cman` to "complete" it.
If someone skips a challenge, it shouldn't show it as completed.

**Comment by Mesure73L** [comment]
If someone marks a challenge as skipped, it will not show as completed. It will just not affect the months or years.

**Comment by BeepBot99** [comment]
I know, but then they will never come back to it. If they just skip it without any status change and go on to future challenges, then sometime later they will come back to it, and they will work hard and complete it, and they will feel really proud.

**Comment by Mesure73L** [comment]
I have an idea:
A way to flag a challenge. This is kinda like my skipping idea, but it will show a flag on the challenge selector for the month, year, and the challenge itself.

**Comment by BeepBot99** [comment]
What does the flag mean?

**Comment by Mesure73L** [comment]
The flag will mean that there is a challenge that you skipped in the year or month, or the challenge itself.

**Comment by BeepBot99** [comment]
Maybe it would work as follows:
1. You hover over the challenge, and on the right side, an uncolored flag shows up
2. You click on the flag, and it becomes colored in.
3. It counts towards completed months and years, but they have a colored-in flag on the right side
4. You can click on the flag on a month or year and a modal will pop up showing which challenges are flagged
5. When you hover over the uncolored flag, there will be a tooltip that says `Postpone Challenge`. When you hover over a colored-in flag, there will be a tooltip that says `Postponed Challenges`.

Is that good? If so, I can make it (because it would require changing the cookies, which I did), or you can make it if you want to.

**Comment by BeepBot99** [comment]
However, this will not be done until 1.1.0, unless we have time earlier.

**Comment by Mesure73L** [comment]
> Maybe it would work as follows:
> 
> 1. You hover over the challenge, and on the right side, an uncolored flag shows up
> 2. You click on the flag, and it becomes colored in.
> 3. It counts towards completed months and years, but they have a colored-in flag on the right side
> 4. You can click on the flag on a month or year and a modal will pop up showing which challenges are flagged
> 5. When you hover over the uncolored flag, there will be a tooltip that says `Postpone Challenge`. When you hover over a colored-in flag, there will be a tooltip that says `Postponed Challenges`.
> 
> Is that good? If so, I can make it (because it would require changing the cookies, which I did), or you can make it if you want to.

Okay, it's good except for step 4. Step 4 is not needed. Just follow the flags until you find the one that's flagged.
